File: src/main/java/ch/uzh/ifi/access/model/Course.java
package ch.uzh.ifi.access.model;

import ch.uzh.ifi.access.model.constants.Visibility;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.Formula;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Getter
@Setter
@Entity
public class Course {
    @Id
    @GeneratedValue
    public Long id;

    @Column(unique = true, nullable = false)
    public String slug;

    @OneToMany(mappedBy = "course", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @MapKey(name = "language")
    public Map<String, CourseInformation> information = new HashMap<>();

    public String logo;

    @Column(nullable = false)
    public String repository;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    public Visibility defaultVisibility;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    public Visibility overrideVisibility;

    @Column(nullable = false)
    public LocalDateTime overrideStart;

    @Column(nullable = false)
    public LocalDateTime overrideEnd;


    // This refers to the role created in keycloak for this course
    @Column(nullable = false)
    public String studentRole;

    @OneToMany(mappedBy = "course", cascade = CascadeType.ALL)
    public List<Assignment> assignments = new ArrayList<>();

    @OneToMany(mappedBy = "course", cascade = CascadeType.ALL)
    public List<Event> events = new ArrayList<>();

    @ElementCollection
    public List<String> supervisors = new ArrayList<>();

    @ElementCollection
    public List<String> assistants = new ArrayList<>();

    @Formula(value = "(SELECT COUNT(*) FROM user_role_mapping u WHERE u.role_id=student_role)")
    public Long studentsCount = 0L;

    @Transient
    public Double points;

    // TODO: why is this here?
    @Transient
    public String userId;

    // TODO: move to service?
    public Assignment createAssignment() {
        Assignment newAssignment = new Assignment();
        assignments.add(newAssignment);
        newAssignment.setCourse(this);
        return newAssignment;
    }
}

File: src/main/java/ch/uzh/ifi/access/model/Course.kt
package ch.uzh.ifi.access.model

import ch.uzh.ifi.access.model.constants.Visibility
import jakarta.persistence.*
import lombok.Getter
import lombok.Setter
import org.hibernate.annotations.Formula
import java.time.LocalDateTime

@Entity
class Course {
    @Id
    @GeneratedValue
    var id: Long? = null

    @Column(unique = true, nullable = false)
    var slug: String? = null

    @OneToMany(mappedBy = "course", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    @MapKey(name = "language")
    var information: MutableMap<String, CourseInformation> = HashMap()
    var logo: String? = null

    @Column(nullable = false)
    var repository: String? = null

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    var defaultVisibility: Visibility? = null

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    var overrideVisibility: Visibility? = null

    @Column(nullable = false)
    var overrideStart: LocalDateTime? = null

    @Column(nullable = false)
    var overrideEnd: LocalDateTime? = null

    // This refers to the role created in keycloak for this course
    @Column(nullable = false)
    var studentRole: String? = null

    @OneToMany(mappedBy = "course", cascade = [CascadeType.ALL])
    var assignments: MutableList<Assignment> = ArrayList()

    @OneToMany(mappedBy = "course", cascade = [CascadeType.ALL])
    var events: MutableList<Event> = ArrayList()

    @ElementCollection
    var supervisors: MutableList<String> = ArrayList()

    @ElementCollection
    var assistants: MutableList<String> = ArrayList()

    @Formula(value = "(SELECT COUNT(*) FROM user_role_mapping u WHERE u.role_id=student_role)")
    var studentsCount = 0L

    @Transient
    var points: Double? = null

    // TODO: why is this here?
    @Transient
    var userId: String? = null

    // TODO: move to service?
    fun createAssignment(): Assignment {
        val newAssignment = Assignment()
        assignments.add(newAssignment)
        newAssignment.course = this
        return newAssignment
    }
}